File: sdk/javascript/src/providers.ts
import fetch from 'node-fetch';

export interface ProviderResult {
  translated_text: string;
  provider: string;
}

export interface TranslationProvider {
  translate(text: string, sourceLang: string, targetLang: string): Promise<ProviderResult>;
  name: string;
}

export class GoogleProvider implements TranslationProvider {
  name = 'Google';

  async translate(text: string, sourceLang: string, targetLang: string): Promise<ProviderResult> {
    const url = `https://translate.googleapis.com/translate_a/single?client=gtx&sl=${sourceLang}&tl=${targetLang}&dt=t&q=${encodeURIComponent(text)}`;
    
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Google Translate HTTP error: ${response.status}`);
    }
    
    const json = await response.json() as any[][];
    
    const translated = json[0]
      ?.map((item: any[]) => item[0])
      .filter(Boolean)
      .join('') || '';
    
    if (!translated) {
      throw new Error('Invalid Google Translate response');
    }
    
    return {
      translated_text: translated,
      provider: this.name,
    };
  }
}

export class BingProvider implements TranslationProvider {
  name = 'Bing';

  async translate(text: string, sourceLang: string, targetLang: string): Promise<ProviderResult> {
    const params = new URLSearchParams();
    params.append('from', sourceLang);
    params.append('to', targetLang);
    params.append('text', text);

    const response = await fetch('https://www.bing.com/ttranslatev3', {
      method: 'POST',
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    if (!response.ok) {
      throw new Error(`Bing Translate HTTP error: ${response.status}`);
    }

    const json = await response.json() as any;
    const translated = json?.[0]?.translations?.[0]?.text;

    if (!translated) {
      throw new Error('Invalid Bing Translate response');
    }

    return {
      translated_text: translated,
      provider: this.name,
    };
  }
}

File: sdk/python/source_translator/providers.py
from abc import ABC, abstractmethod
from typing import Optional
import httpx
from pydantic import BaseModel


class ProviderResult(BaseModel):
    translated_text: str
    provider: str


class TranslationProvider(ABC):
    @abstractmethod
    async def translate(self, text: str, source_lang: str, target_lang: str) -> ProviderResult:
        pass
    
    @property
    @abstractmethod
    def name(self) -> str:
        pass


class GoogleProvider(TranslationProvider):
    name = "Google"

    async def translate(self, text: str, source_lang: str, target_lang: str) -> ProviderResult:
        url = f"https://translate.googleapis.com/translate_a/single?client=gtx&sl={source_lang}&tl={target_lang}&dt=t&q={text}"
        
        async with httpx.AsyncClient() as client:
            response = await client.get(url)
            response.raise_for_status()
            data = response.json()
        
        translated = "".join(
            item[0] for item in data[0] if item[0]
        )
        
        if not translated:
            raise ValueError("Invalid Google Translate response")
        
        return ProviderResult(
            translated_text=translated,
            provider=self.name,
        )


class BingProvider(TranslationProvider):
    name = "Bing"

    async def translate(self, text: str, source_lang: str, target_lang: str) -> ProviderResult:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                "https://www.bing.com/ttranslatev3",
                data={
                    "from": source_lang,
                    "to": target_lang,
                    "text": text,
                },
                headers={
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
                },
            )
            response.raise_for_status()
            data = response.json()
        
        translated = data[0]["translations"][0]["text"]
        
        if not translated:
            raise ValueError("Invalid Bing Translate response")
        
        return ProviderResult(
            translated_text=translated,
            provider=self.name,
        )
